### PR TITLE
Code format

### PR DESCRIPTION
**Pull request by janniklasrose**
I've applied `black` to both the source files in `chebpy/` and the tests in `tests/`.

I'm happy with what it does to the source, but not so regarding the tests. I think it's probably fair to revert 69d4c206ec1a6b5f263846bb23ca42d63b1f2e9d and exclude auto-formatting from the test files, where non-compliant formatting is frequently used for legibility.

### CODE REVIEW

File: tests/test_bndfun.py [hunk 1]
```diff
@@ -95,72 +96,79 @@ def test_call_clenshaw(self):
...
-    (lambda x: 1e10*exp(x),      [-1,1],      1e10*exp(1)),
-    (lambda x: 0*x+1.,           [-1e5,1e4],  1),
+    (lambda x: sin(4 * pi * x), [-2, 2], 1),
+    (lambda x: cos(x), [-10, 1], 1),
+    (lambda x: cos(4 * pi * x), [-100, 100], 1),
+    (lambda x: exp(cos(4 * pi * x)), [-1, 1], exp(1)),
+    (lambda x: cos(3244 * x), [-2, 0], 1),
+    (lambda x: exp(x), [-1, 2], exp(2)),
+    (lambda x: 1e10 * exp(x), [-1, 1], 1e10 * exp(1)),
+    (lambda x: 0 * x + 1.0, [-1e5, 1e4], 1),
```
Comment (janniklasrose) [inline tests/test_bndfun.py:157]: This is one example of where the auto-formatter makes the set of test functions hard to read

File: tests/test_bndfun.py [hunk 2]
```diff
@@ -219,30 +228,35 @@ def test_cumsum_empty(self):
...
-    (lambda x: 0*x+1.,           [-100,300],                   400,      eps),
+    (lambda x: sin(x), [-2, 2], 0.0, 2 * eps),
+    (lambda x: sin(4 * pi * x), [-0.1, 0.7], 0.088970317927147, 1e1 * eps),
+    (lambda x: cos(x), [-100, 203], 0.426944059057085, 4e2 * eps),
+    (lambda x: cos(4 * pi * x), [-1e-1, -1e-3], 0.074682699182803, 2 * eps),
+    (lambda x: exp(cos(4 * pi * x)), [-3, 1], 5.064263511008033, 4 * eps),
+    (lambda x: cos(3244 * x), [0, 0.4], -3.758628487169980e-05, 5e2 * eps),
+    (lambda x: exp(x), [-2, -1], exp(-1) - exp(-2), 2 * eps),
+    (lambda x: 1e10 * exp(x), [-1, 2], 1e10 * (exp(2) - exp(-1)), 2e10 * eps),
+    (lambda x: 0 * x + 1.0, [-100, 300], 400, eps),
```
Comment (janniklasrose) [inline tests/test_bndfun.py:245]: Another example

File: tests/test_bndfun.py [hunk 3]
```diff
@@ -253,29 +267,33 @@ def tester(self):
...
+    (lambda x: 0 * x + 1.0, lambda x: x, [-2, 3], eps),
+    (lambda x: x, lambda x: 1 / 2 * x ** 2, [-5, 0], 4 * eps),
+    (lambda x: x ** 2, lambda x: 1 / 3 * x ** 3, [1, 10], 2e2 * eps),
+    (lambda x: x ** 3, lambda x: 1 / 4 * x ** 4, [-1e-2, 4e-1], 2 * eps),
+    (lambda x: x ** 4, lambda x: 1 / 5 * x ** 5, [-3, -2], 3e2 * eps),
+    (lambda x: x ** 5, lambda x: 1 / 6 * x ** 6, [-1e-10, 1], 4 * eps),
+    (lambda x: sin(x), lambda x: -cos(x), [-10, 22], 3e1 * eps),
+    (lambda x: cos(3 * x), lambda x: 1.0 / 3 * sin(3 * x), [-3, 4], 2 * eps),
+    (lambda x: exp(x), lambda x: exp(x), [-60, 1], 1e1 * eps),
+    (lambda x: 1e10 * exp(x), lambda x: 1e10 * exp(x), [-1, 1], 1e10 * (3 * eps)),
```
Comment (janniklasrose) [inline tests/test_bndfun.py:279]: Another

File: tests/test_bndfun.py [hunk 4]
```diff
@@ -285,65 +303,70 @@ def tester(self):
...
+    (lambda x: 0 * x + 1.0, lambda x: 0 * x + 0, [-2, 3], eps),
+    (lambda x: x, lambda x: 0 * x + 1, [-5, 0], 2e1 * eps),
+    (lambda x: x ** 2, lambda x: 2 * x, [1, 10], 2e2 * eps),
+    (lambda x: x ** 3, lambda x: 3 * x ** 2, [-1e-2, 4e-1], 3 * eps),
+    (lambda x: x ** 4, lambda x: 4 * x ** 3, [-3, -2], 1e3 * eps),
+    (lambda x: x ** 5, lambda x: 5 * x ** 4, [-1e-10, 1], 4e1 * eps),
+    (lambda x: sin(x), lambda x: cos(x), [-10, 22], 5e2 * eps),
+    (lambda x: cos(3 * x), lambda x: -3 * sin(3 * x), [-3, 4], 5e2 * eps),
+    (lambda x: exp(x), lambda x: exp(x), [-60, 1], 2e2 * eps),
+    (lambda x: 1e10 * exp(x), lambda x: 1e10 * exp(x), [-1, 1], 1e10 * 2e2 * eps),
```
Comment (janniklasrose) [inline tests/test_bndfun.py:316]: Another

File: tests/test_bndfun.py [hunk 5]
```diff
@@ -367,54 +390,61 @@ def test_onefun_construction(self):
...
-    (lambda x: sin(x),              "sin(x)",   [-2,3], 20),
-    (lambda x: cos(20*x),           "cos(20x)", [-2,3], 90),
-    (lambda x: 0.*x+1.,             "constfun", [-2,3],  1),
-    (lambda x: 0.*x,                "zerofun",  [-2,3],  1),
+    (lambda x: x ** 3 + x ** 2 + x + 1, "poly3(x)", [-2, 3], 4),
+    (lambda x: exp(x), "exp(x)", [-2, 3], 20),
+    (lambda x: sin(x), "sin(x)", [-2, 3], 20),
+    (lambda x: cos(20 * x), "cos(20x)", [-2, 3], 90),
+    (lambda x: 0.0 * x + 1.0, "constfun", [-2, 3], 1),
+    (lambda x: 0.0 * x, "zerofun", [-2, 3], 1),
```
Comment (janniklasrose) [inline tests/test_bndfun.py:447]: Another

File: tests/test_bndfun.py [hunk 6]
```diff
@@ -692,102 +750,297 @@ def tester(self):
...
+        [
+            ([uf3, (-0.8, 0.8)], eps),
+        ],
+    ),
+    (
+        np.tanh,
+        [
+            ([uf3, (-3, 3)], eps),
+        ],
+    ),
```
Comment (janniklasrose) [inline tests/test_bndfun.py:980]: This one is particularly bad

File: tests/test_bndfun.py [hunk 7]
```diff
@@ -692,102 +750,297 @@ def tester(self):
...
     return tester
 
-for (ufunc,  [([f, intvl], tol), ]) in ufunc_test_params:
+
+for (
+    ufunc,
+    [
+        ([f, intvl], tol),
+    ],
+) in ufunc_test_params:
```
Comment (janniklasrose) [inline tests/test_bndfun.py:1003]: Not pretty

File: tests/test_bndfun.py [hunk 8]
```diff
@@ -692,102 +750,297 @@ def tester(self):
...
-    (lambda x: sin(100*pi*x), [-0.5,0.5], np.linspace(-.5,.5,101),           eps),
-    (lambda x: sin(5*pi/2*x), [-1,1],     np.array([-.8, -.4, 0, .4, .8]),   eps)
-    )
+    (lambda x: 3 * x + 2.0, [-2, 3], np.array([-2 / 3]), eps),
+    (lambda x: x ** 2 + 0.2 * x - 0.08, [-2, 5], np.array([-0.4, 0.2]), 3e1 * eps),
+    (lambda x: sin(x), [-7, 7], pi * np.linspace(-2, 2, 5), 1e1 * eps),
+    (lambda x: cos(2 * pi * x), [-20, 10], np.linspace(-19.75, 9.75, 60), 3e1 * eps),
+    (lambda x: sin(100 * pi * x), [-0.5, 0.5], np.linspace(-0.5, 0.5, 101), eps),
+    (lambda x: sin(5 * pi / 2 * x), [-1, 1], np.array([-0.8, -0.4, 0, 0.4, 0.8]), eps),
+)
```
Comment (janniklasrose) [inline tests/test_bndfun.py:1043]: Last example in this file

File: tests/test_chebfun.py [hunk 9]
```diff
@@ -74,57 +78,60 @@ def test_initempty(self):
...
         a, b = ff.breakdata.keys()
-        fa, fb, = ff.breakdata.values()
-        self.assertEqual(a,-2)
-        self.assertEqual(b,-1)
-        self.assertLessEqual(abs(fa-self.f(-2)), eps)
-        self.assertLessEqual(abs(fb-self.f(-1)), eps)
+        (
+            fa,
+            fb,
+        ) = ff.breakdata.values()
```
Comment (janniklasrose) [inline tests/test_chebfun.py:118]: remove the trailing comma after fb to format in-line again

File: tests/test_chebfun.py [hunk 10]
```diff
@@ -137,46 +144,49 @@ def test_initfun_adaptive_empty_domain(self):
...
```
Comment (janniklasrose) [inline tests/test_chebfun.py:153]: trailing comma

File: tests/test_chebfun.py [hunk 11]
```diff
@@ -697,86 +729,286 @@ def tester(self):
...
+        [
+            ([uf3, (-0.8, 0.8)], eps),
+        ],
+    ),
+    (
+        np.tanh,
+        [
+            ([uf3, (-3, 3)], eps),
+        ],
+    ),
```
Comment (janniklasrose) [inline tests/test_chebfun.py:971]: Also very bad

File: tests/test_chebtech.py [hunk 12]
```diff
@@ -83,20 +87,24 @@ def asserter(self):
...
-    (Chebtech2._chebpts(5), np.array([-1., -2.**(-.5), 0., 2.**(-.5), 1.]), eps),
+    (Chebtech2._chebpts(1), np.array([0.0]), eps),
+    (Chebtech2._chebpts(2), np.array([-1.0, 1.0]), eps),
+    (Chebtech2._chebpts(3), np.array([-1.0, 0.0, 1.0]), eps),
+    (Chebtech2._chebpts(4), np.array([-1.0, -0.5, 0.5, 1.0]), 2 * eps),
+    (
+        Chebtech2._chebpts(5),
+        np.array([-1.0, -(2.0 ** (-0.5)), 0.0, 2.0 ** (-0.5), 1.0]),
+        eps,
+    ),
```
Comment (janniklasrose) [inline tests/test_chebtech.py:103]: Possibly okay, still legible

File: tests/test_chebtech.py [hunk 13]
```diff
@@ -177,54 +187,59 @@ def test_call_raises(self):
...
-    (lambda x: 1e10*exp(x),        15, 1e10*exp(1)),
-    (lambda x: 0*x+1.,              1, 1),
+    (lambda x: sin(4 * pi * x), 40, 1),
+    (lambda x: cos(x), 15, 1),
+    (lambda x: cos(4 * pi * x), 39, 1),
+    (lambda x: exp(cos(4 * pi * x)), 181, exp(1)),
+    (lambda x: cos(3244 * x), 3389, 1),
+    (lambda x: exp(x), 15, exp(1)),
+    (lambda x: 1e10 * exp(x), 15, 1e10 * exp(1)),
+    (lambda x: 0 * x + 1.0, 1, 1),
```
Comment (janniklasrose) [inline tests/test_chebtech.py:229]: Maybe we can just use format fences to disable autoformat on these examples

File: tests/test_chebtech.py [hunk 14]
```diff
@@ -319,35 +334,39 @@ def test_real_imag(self):
...
-    (lambda x: 0*x+1.,              1,                     2,      eps),
+    (lambda x: sin(x), 14, 0.0, eps),
+    (lambda x: sin(4 * pi * x), 40, 0.0, 1e1 * eps),
+    (lambda x: cos(x), 15, 1.682941969615793, 2 * eps),
+    (lambda x: cos(4 * pi * x), 39, 0.0, 2 * eps),
+    (lambda x: exp(cos(4 * pi * x)), 182, 2.532131755504016, 4 * eps),
+    (lambda x: cos(3244 * x), 3389, 5.879599674161602e-04, 5e2 * eps),
+    (lambda x: exp(x), 15, exp(1) - exp(-1), 2 * eps),
+    (lambda x: 1e10 * exp(x), 15, 1e10 * (exp(1) - exp(-1)), 4e10 * eps),
+    (lambda x: 0 * x + 1.0, 1, 2, eps),
```
Comment (janniklasrose) [inline tests/test_chebtech.py:356]: here, too

File: tests/test_chebtech.py [hunk 15]
```diff
@@ -358,28 +377,32 @@ def tester(self):
...
+    (lambda x: 0 * x + 1.0, lambda x: x, 1, eps),
+    (lambda x: x, lambda x: 1 / 2 * x ** 2, 2, 2 * eps),
+    (lambda x: x ** 2, lambda x: 1 / 3 * x ** 3, 3, 2 * eps),
+    (lambda x: x ** 3, lambda x: 1 / 4 * x ** 4, 4, 2 * eps),
+    (lambda x: x ** 4, lambda x: 1 / 5 * x ** 5, 5, 2 * eps),
+    (lambda x: x ** 5, lambda x: 1 / 6 * x ** 6, 6, 4 * eps),
+    (lambda x: sin(x), lambda x: -cos(x), 16, 2 * eps),
+    (lambda x: cos(3 * x), lambda x: 1.0 / 3 * sin(3 * x), 23, 2 * eps),
+    (lambda x: exp(x), lambda x: exp(x), 16, 3 * eps),
+    (lambda x: 1e10 * exp(x), lambda x: 1e10 * exp(x), 16, 1e10 * (3 * eps)),
```
Comment (janniklasrose) [inline tests/test_chebtech.py:389]: and here

File: tests/test_chebtech.py [hunk 16]
```diff
@@ -390,26 +413,30 @@ def tester(self):
...
+    (lambda x: 0 * x + 1.0, lambda x: 0 * x + 0, 1, eps),
+    (lambda x: x, lambda x: 0 * x + 1, 2, 2 * eps),
+    (lambda x: x ** 2, lambda x: 2 * x, 3, 2 * eps),
+    (lambda x: x ** 3, lambda x: 3 * x ** 2, 4, 2 * eps),
+    (lambda x: x ** 4, lambda x: 4 * x ** 3, 5, 3 * eps),
+    (lambda x: x ** 5, lambda x: 5 * x ** 4, 6, 4 * eps),
+    (lambda x: sin(x), lambda x: cos(x), 16, 5e1 * eps),
+    (lambda x: cos(3 * x), lambda x: -3 * sin(3 * x), 23, 5e2 * eps),
+    (lambda x: exp(x), lambda x: exp(x), 16, 2e2 * eps),
+    (lambda x: 1e10 * exp(x), lambda x: 1e10 * exp(x), 16, 1e10 * 2e2 * eps),
```
Comment (janniklasrose) [inline tests/test_chebtech.py:425]: and here

File: tests/test_chebtech.py [hunk 17]
```diff
@@ -610,124 +644,133 @@ def test__neg__empty(self):
...
-    (lambda x: sin(5*pi/2*x), np.array([-.8, -.4, 0, .4, .8]),        1*eps)
-    )
+    (lambda x: 3 * x + 2.0, np.array([-2 / 3]), 1 * eps),
+    (lambda x: x ** 2, np.array([0.0, 0.0]), 1 * eps),
+    (lambda x: x ** 2 + 0.2 * x - 0.08, np.array([-0.4, 0.2]), 1 * eps),
+    (lambda x: sin(x), np.array([0]), 1 * eps),
+    (lambda x: cos(2 * pi * x), np.array([-0.75, -0.25, 0.25, 0.75]), 1 * eps),
+    (lambda x: sin(100 * pi * x), np.linspace(-1, 1, 201), 1 * eps),
+    (lambda x: sin(5 * pi / 2 * x), np.array([-0.8, -0.4, 0, 0.4, 0.8]), 1 * eps),
+)
```
Comment (janniklasrose) [inline tests/test_chebtech.py:773]: last one in this file

File: tests/utilities.py [hunk 18]
```diff
@@ -24,13 +29,13 @@ def asserter(self):
...
-    (lambda x: np.cos(20*x),          'cos(20x)',       51, True),
-    (lambda x: 0.*x+1.,               'constfun',        1, False),
-    (lambda x: 0.*x,                  'zerofun',         1, True),
+    (lambda x: x ** 3 + x ** 2 + x + 1.1, "poly3(x)", 4, True),
+    (lambda x: np.exp(x), "exp(x)", 15, False),
+    (lambda x: np.sin(x), "sin(x)", 14, True),
+    (lambda x: 0.2 + 0.1 * np.sin(x), "(.2+.1*sin(x))", 14, False),
+    (lambda x: np.cos(20 * x), "cos(20x)", 51, True),
+    (lambda x: 0.0 * x + 1.0, "constfun", 1, False),
+    (lambda x: 0.0 * x, "zerofun", 1, True),
```
Comment (janniklasrose) [inline tests/utilities.py:38]: Only one in this file

**Review by janniklasrose — COMMENTED** [review]
I've skimmed through all the changes with particular focus on format changes that make sets of test functions/conditions harder to identify. I think overall `tests/` can also benefit from black, but maybe a few `format: on/off` fences are needed to make the code more legible

**Comment by janniklasrose** [comment]
What's your view regarding `tests/` and the resulting `black` style overall? If you're okay with it, I'll include a CI workflow to his PR that auto-formats the code on `master`. I think there is a way to tell `git` through commit naming to ignore the auto-formatting commits in `git blame` but I also remember reading that GitHub does not support this (yet)

**Comment by markrichardson** [comment]
Interesting. This is largely fine with me - even the behaviour in `tests/`. It's true that I set up the alignment in `tests/` previously in order to improve readability, but I'm not sure what `black` has done has necessarily degraded the experience. In fact it probably makes things better.

**Comment by janniklasrose** [comment]
Actually, one could argue that code such as the one you commented on (and some other bits in my review, too) was not good to begin with and as a result `black`'s formatting is bad.

For instance, the particularly bad example above could (probably should) be refactored to avoid a lot of the repetition: `uf1`, `uf2`, `uf3` have different test blocks and every call uses `eps`.

**Comment by markrichardson** [comment]
yeah that's fair

**Comment by janniklasrose** [comment]
I'll proceed with adding the CI action - which one do you prefer: auto-formatter or format-checker?

**Comment by markrichardson** [comment]
I'm kind of inclined to let Black do its thing in full (for ease) [i.e. auto-formatter] and then we can rework any parts of it that stand out painfully later?

**Comment by markrichardson** [comment]
hey Jan - apologies if this was on me to merge. good to do so from your perspective?

**Comment by janniklasrose** [comment]
Hi Mark, not yet. I've been busy with other stuff, apologies. I shall have a look this weekend, still need to add the GitHub Action. I did some digging and the consensus seems to be to not auto-format as part of the automatic workflows since this can have unintended side effects. Better to have a linter only and thus force contributors to apply black themselves.

**Comment by markrichardson** [comment]
roger that. will sit tight - thanks for the update

**Comment by janniklasrose** [comment]
Should add a reference to `black` in the developer notes (#55)

**Comment by janniklasrose** [comment]
Ready for merging now.

I'm also considering adding `flake8` as a linter, but that's something for a later PR. This one changes almost every file and flake8 compliance will again make a lot of changes, so that should be in separate PRs for more sane reviewing.

File: .github/workflows/format.yml [hunk 19]
```diff
@@ -0,0 +1,18 @@
...
+jobs:
+
+  black-check:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - name: Set up Python
+        uses: actions/setup-python@v2
+        with:
+          python-version: 3.8
```
Comment (janniklasrose) [inline .github/workflows/format.yml:14]: Inconsistent with `target-version = ['py37']` in pyproject.toml

File: pyproject.toml [hunk 20]
```diff
@@ -21,3 +21,8 @@ report.exclude_lines = [
     # we don't want to test this, usually abc methods
     "raise NotImplementedError",
 ]
+
+# Black style
+[tool.black]
+
+target-version = ['py37']
```
Comment (janniklasrose) [inline pyproject.toml:28]: Python 3.7 is the lowest version supported in chebpy, so I chose that

**Comment by markrichardson** [comment]
perfect, merging now

**Pull request merged**